library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java

```diff
@@ -0,0 +1,234 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.lib.util;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datatorrent.netlet.util.DTThrowable;
+
+/**
+ * Abstract implementation of {@link AppIdAwareStorageAgent} which can be
+ * configured be KeyValue store witch implementation of {@link KeyValueStore}
+ * 
+ * NOTE - this should be picked from APEX-CORE once below feature is release
+ * https://issues.apache.org/jira/browse/APEXCORE-283
+ * 
+ * @param <S>
+ *          Store implementation
+ */
+public abstract class AbstractKeyValueStorageAgent<S extends StorageAgentKeyValueStore>
+    implements AppIdAwareStorageAgent, Serializable
+{
+
+  protected S store;
+  protected String applicationId;
+  public static final String CHECKPOINT_KEY_SEPARATOR = "-";
+
+  /**
+   * Gets the store
+   *
+   * @return the store
+   */
+  public S getStore()
+  {
+    return store;
+  }
+
+  /**
+   * Sets the store
+   *
+   * @param store
+   */
+  public void setStore(S store)
+  {
+    this.store = store;
+  }
+
+  /**
+   * Return yarn application id of running application
+   * 
+   * @return
+   */
+  public String getApplicationId()
+  {
+    return applicationId;
+  }
+
+  /**
+   * Set yarn application id
+   * 
+   * @param applicationId
+   */
+  public void setApplicationId(String applicationId)
+  {
+    this.applicationId = applicationId;
+  }
+
+  /**
+   * Generates key from operator id and window id to store unique operator
+   * checkpoints
+   * 
+   * @param operatorId
+   * @param windowId
+   * @return unique key for store
+   */
+  public static String generateKey(int operatorId, long windowId)
+  {
+    return String.valueOf(operatorId) + CHECKPOINT_KEY_SEPARATOR + String.valueOf(windowId);
+  }
+
+  /**
+   * Stores the given operator object in configured store
+   * 
+   * @param object
+   *          Operator object to store
+   * @param operatorId
+   *          of operator
+   * @param windowId
+   *          window id of operator to checkpoint
+   * 
+   */
+  @Override
+  public void save(Object object, int operatorId, long windowId) throws IOException
+  {
+
+    try {
+      store(generateKey(operatorId, windowId), object);
+      logger.info("saved check point object key {} region {}", generateKey(operatorId, windowId), applicationId);
+    } catch (Throwable t) {
```

Comment:
Initial though was, since this is abstract implementation and to ensure that any errors or exceptions thrown by underlying store should not cause disruption to application.

I can change this to catch to run time exception instead.